### PR TITLE
Dev refactor onediff comfy nodes

### PR DESCRIPTION
**Pull request by ccssu**
### Install

1. [OneDiff Installation Guide](https://github.com/siliconflow/onediff/blob/main/README_ENTERPRISE.md#install-onediff-enterprise)
2. [OneDiffx Installation Guide](https://github.com/siliconflow/onediff/tree/main/onediff_diffusers_extensions#install-and-setup)

## Usage:
| Option                                 | Range  | Default | Description                                                                  |
| -------------------------------------- | ------ | ------- | ---------------------------------------------------------------------------- |
| quantized_conv_percentage                | [0, 100] | 100     |  Example value representing 100% quantization for linear layers｜
| quantized_linear_percentage           | [0, 100] | 100     | Example value representing 100% quantization for convolutional layers  |
| conv_compute_density_threshold    | [0, ∞) | 100     | Computational density threshold for quantizing convolutional modules to 100. |
| linear_compute_density_threshold  | [0, ∞) | 300     | Computational density threshold for quantizing linear modules to 300.        |

Notes:

1. Specify the directory for saving graphsusing export COMFYUI_ONEDIFF_SAVE_GRAPH_DIR="/path/to/save/graphs".
2. The log *.pt file is cached. Quantization result information can be found in `cache_dir`/quantization_stats.json.

### [SDXL](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0)
Updated on Mon 08 Apr 2024 

quant_sdxl

| Accelerator             | Baseline (non-optimized) | OneDiff(optimized) | OneDiff Quant(optimized) |
| ----------------------- | ------------------------ | ------------------ | ------------------------ |
| NVIDIA GeForce RTX 3090 | 8.03 s                   | 4.44 s ( ~44.7%)   | 3.34 s ( ~58.4%)         |

- torch   `python -c "import torch; print(torch.__version__)"`: {version: 2.2.1+cu121}
- oneflow  `python -m oneflow --doctor`: {git_commit: 710818c, version: 0.9.1.dev20240406+cu121, enterprise: True}
- Start comfyui command: `python main.py --gpu-only`

### SD1.5

sd 1.5

### SVD
svd_quant

### CODE REVIEW

File: onediff_comfy_nodes/__init__.py [hunk 1]
```diff
@@ -1,84 +1,30 @@
...
-            "OneDiffQuantCheckpointLoaderSimple": OneDiffQuantCheckpointLoaderSimple,
-            "OneDiffQuantCheckpointLoaderSimpleAdvanced": OneDiffQuantCheckpointLoaderSimpleAdvanced,
-            "ImageOnlyOneDiffQuantCheckpointLoaderAdvanced": ImageOnlyOneDiffQuantCheckpointLoaderAdvanced,
-            "QuantKSampler": QuantKSampler,
-        }
-    )
+extras = [nodes_compare, nodes_oneflow_optimizer, nodes_torch_compile_optimizer]
+for node in extras:
+    NODE_CLASS_MAPPINGS.update(node.NODE_CLASS_MAPPINGS)
+    NODE_DISPLAY_NAME_MAPPINGS.update(node.NODE_DISPLAY_NAME_MAPPINGS)
```
Comment (ccssu) [inline onediff_comfy_nodes/__init__.py:29]: 参考 comfyui 做了隔离，新的模块往 extras_node 上面堆。
image

**Comment by ccssu** [comment]
| Accelerator             | Baseline (non-optimized) | OneDiff(optimized) | OneDiff Quant(optimized) |
| ----------------------- | ------------------------ | ------------------ | ------------------------ |
| NVIDIA A800-SXM4-80GB | 63.64 s                   | 47.57 s ( ~25.25%)   | todo         |

The following table shows the workflows used separately：
  | Baseline (non-optimized) | OneDiff(optimized) | OneDiff Quant(optimized) |
 | ------------------------ | ------------------ | ------------------------ |
|svd_ baseline|OneDiff(optimized)|OneDiff Quant(optimized)|

File: onediff_comfy_nodes/_config.py [hunk 2]
```diff
@@ -1,44 +1,6 @@
...
-
-    opt_models_dir = Path(models_dir) / ONEDIFF_QUANTIZED_OPTIMIZED_MODELS
-    opt_models_dir.mkdir(parents=True, exist_ok=True)
-
-    folder_names_and_paths[ONEDIFF_QUANTIZED_OPTIMIZED_MODELS] = (
-        [str(opt_models_dir)],
-        supported_pt_extensions,
-    )
+# TODO fix path
+os.environ["COMFYUI_ROOT"] = "/home/fengwen/workspace/ComfyUI"
```
Comment (lixiang007666) [inline onediff_comfy_nodes/_config.py:6]: 这个 path 还没修改。

File: onediff_comfy_nodes/modules/oneflow/hijack_animatediff/_config.py [hunk 3]
```diff
@@ -3,9 +3,12 @@
 commit: 5d875d77fe6e31a4b0bc6dc36f0441eba3b6afe1 
 """
 import os
-from ..sd_hijack_utils import Hijacker
-from onediff.infer_compiler.transform import transform_mgr
+
 from onediff.infer_compiler.import_tools import DynamicModuleLoader
+from onediff.infer_compiler.transform import transform_mgr
+
+from ...sd_hijack_utils import Hijacker
```
[the remark this answers is not in the document]
Comment (ccssu) [reply onediff_comfy_nodes/modules/oneflow/hijack_animatediff/_config.py:10]: 主要是考虑安全，使用相对导入避免全局命名冲突。
 image

`from sd_hijack_utils import Hijacker` 要这样导入 需要单独放个目录 然后放 `sd_hijack_utils.py` 才安全 。

File: onediff_comfy_nodes/_config.py (hunk 2 above)
Comment (ccssu) [reply onediff_comfy_nodes/_config.py]: 已修复

File: onediff_comfy_nodes/modules/oneflow/utils/optimization_tools.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
+import torch
```
Comment (doombeaker) [inline onediff_comfy_nodes/modules/oneflow/utils/optimization_tools.py:1]: 可能改成 xxx_utils.py 比较好

File: onediff_comfy_nodes/extras_nodes/nodes_oneflow_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,718 @@
...
+        )
+        op = OptimizerScheduler(DeepcacheOptimizerExecutor(
+               cache_interval=cache_interval,
+                cache_layer_id=cache_layer_id,
+                cache_block_id=cache_block_id,
+                start_step=start_step,
+                end_step=end_step,
+        ))
+
+        modelpatcher = op.compile(modelpatcher, ckpt_name=ckpt_name)
```
Comment (doombeaker) [inline onediff_comfy_nodes/extras_nodes/nodes_oneflow_optimizer.py:313]:
```suggestion
        modelpatcher = OptimizerScheduler(DeepcacheOptimizerExecutor(
               cache_interval=cache_interval,
                cache_layer_id=cache_layer_id,
                cache_block_id=cache_block_id,
                start_step=start_step,
                end_step=end_step,
        ))(modelpatcher, ckpt_name=ckpt_name)
```

我看 scheduler 里重写了 `__call__` ，但是所有的 scheuler 调用都还是用的 compile，这是为什么呢。

或者说为什么重写 `__call__`，但是不倾向于使用它呢。

File: onediff_comfy_nodes/extras_nodes/nodes_oneflow_optimizer.py [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+
+    @torch.no_grad()
+    def speedup(self, model, inplace=False, cache_name = "svd", custom_optimizer: OptimizerScheduler=None):
+        if custom_optimizer:
+            op = custom_optimizer
+            op.inplace = inplace
+        else:
+            op = OptimizerScheduler(BasicOneFlowOptimizerExecutor(), inplace=inplace)
+
+        optimized_model = op.compile(model, ckpt_name=cache_name)
```
Comment (doombeaker) [inline onediff_comfy_nodes/extras_nodes/nodes_oneflow_optimizer.py:370]:
```suggestion
        optimized_model = op(model, ckpt_name=cache_name)
```

File: onediff_comfy_nodes/modules/oneflow/utils/optimization_tools.py (hunk 4 above)
Comment (ccssu) [reply onediff_comfy_nodes/modules/oneflow/utils/optimization_tools.py]: 已修复为 onediff_comfy_nodes/modules/oneflow/utils/booster_utils.py

File: onediff_comfy_nodes/modules/optimizer_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+from typing import List
+
+from comfy.model_patcher import ModelPatcher
+from onediff.infer_compiler.oneflow import \
+    OneflowDeployableModule as DeployableModule
+
+from .optimizer_interface import OptimizerExecutor
+
+
+class OptimizerScheduler:
```
[the remark this answers is not in the document]
Comment (ccssu) [reply onediff_comfy_nodes/modules/optimizer_scheduler.py]: 已修复 替换 OptimizerXXXX 为 BoosterXXXX

File: onediff_comfy_nodes/extras_nodes/nodes_oneflow_optimizer.py (hunk 5 above)
Comment (ccssu) [reply onediff_comfy_nodes/extras_nodes/nodes_oneflow_optimizer.py]: 当时想的是体现 compile， 已修复为统一用 `__call__ `调用

File: onediff_comfy_nodes/modules/optimizer_scheduler.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hjchen2) [reply onediff_comfy_nodes/modules/optimizer_scheduler.py]: 不如就叫ExecutorScheduler，OptimizerExecutor就简化为Executor

**Review by doombeaker — APPROVED** [review]

File: onediff_comfy_nodes/modules/optimizer_scheduler.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ccssu) [reply onediff_comfy_nodes/modules/optimizer_scheduler.py]: 这个 **Booster** 主要在于 cmfyui ui 界面体现 新意 。
image
